langchain/llms/databricks_endpoint.py

```diff
@@ -0,0 +1,224 @@
+from abc import ABC, abstractmethod
+import logging
+import os
+from typing import Any, Dict, List, Mapping, Optional, Union, Callable
+
+from pydantic import BaseModel, Extra, Field, validator, root_validator, ValidationError, PrivateAttr
+import requests
+
+from langchain.llms.base import LLM
+from langchain.llms.utils import enforce_stop_tokens
+
+__all__ = ["DatabricksEndpoint"]
+
+
+class DatabricksClientBase(BaseModel, ABC):
+    api_url: str
+    api_token: str
+
+    def post_raw(self, request):
+        headers = {"Authorization": f"Bearer {self.api_token}"}
+        response = requests.post(self.api_url, headers=headers, json=request)
+        # TODO: error handling and automatic retries
+        assert response.ok, f"HTTP {response.status_code} error: {response.text}"
+        return response.json()
+    
+    @abstractmethod
+    def post(self, request):
+        raise NotImplementedError
+
+
+class DatabricksServingEndpointClient(DatabricksClientBase):
+    host: str
+    endpoint_name: str
+
+    @root_validator(pre=True)
+    def set_api_url(cls, values):
+        if "api_url" not in values:
+            values["api_url"] = \
+                f"https://{values['host']}/serving-endpoints/{values['endpoint_name']}/invocations"
+        return values
+
+    def post(self, request):
+        # See https://docs.databricks.com/machine-learning/model-serving/score-model-serving-endpoints.html
+        wrapped_request = {"dataframe_records": [request]}
+        wrapped_response = self.post_raw(wrapped_request)
+        return wrapped_response["predictions"][0]
+
+class DatabricksClusterDriverProxyClient(DatabricksClientBase):
+    host: str
+    cluster_id: str
+    cluster_driver_port: str
+
+    @root_validator(pre=True)
+    def set_api_url(cls, values):
+        if "api_url" not in values:
+            values["api_url"] = \
+                f"https://{values['host']}/driver-proxy-api/o/0/{values['cluster_id']}/{values['cluster_driver_port']}"
+        return values
+
+    def post(self, request):
+        return self.post_raw(request)
+
+def get_repl_context():
+    """Gets the notebook REPL context if running inside a Databricks notebook.
+    Returns None otherwise.
+    """
+    context = None
+    try:
+        from dbruntime.databricks_repl_context import get_context
+        context = get_context()
+    except ImportError:
+        logging.debug("Not running inside a Databricks notebook")
+    except Exception:
+        logging.debug(
+            "Error getting the REPL context. Will not use it to provide default values.",
+            exc_info=True)
+    return context
+
+def get_default_host():
```

Comment:
I will make a follow-up to create Databricks utility methods for getting the host and api_token.